Fix #3295 Save draft change list to local storage.

Anthony-Alridge opened the pull request:
Implements Milestone 3 of the task to extend the changeLists autosave functionality to local saves. 

A couple of preemptive questions about this PR:
  - Testing: I wasn't entirely sure how to go about unit testing (or if unit testing is applicable). I thought pushing might help to clarify why but the changes are around the functionality in exploration data which is not tested directly (it is replicated using mock objects in tests). I think that an end to end test is more suitable to test the changes which won't be applicable until completion of the next milestone (I did manually test this PR to make sure it is behaving as expected).
 - If local storage doesn't exist: as discussed in the design doc if local storage doesn't exist then on the error callback a message should be displayed to the user (in the case of an internet connection issue) with the changes they may lose as opposed the current system in which no warning is currently displayed (in fact it seems it no longer warns the user they are offline and to refresh). To implement this I was thinking using a modal and checking if local storage doesn't exist in which case if this isn't a validation error then we display the modal. However as we are offline the modal template url won't be retrievable so it would have to be pre-loaded for this to work. This seems like a relatively large change so I thought perhaps if we wish to display the message it may be worth making another milestone in the design doc and committing this as the next PR. What do you think?

Commit:
add change list id field to user data model and update tests

Commit:
edit create or update draft to increment change list id(#3295)

Commit:
fix syntax error in test case

Commit:
fix lint error in test case

Commit:
Fix small spacing error

Commit:
remove trailing whitespace

Commit:
return draft id with exploration data and autosave put and add tests

Commit:
add draft id to exploration data factory method

Commit:
Merge branch 'develop' into autosave-local

Commit:
fix indentation error

Commit:
address review comments (#3565)

Commit:
fix indentation error

Commit:
Address review comments (#3565)

Commit:
Merge branch 'develop' into autosave-local

Commit:
Address review comments (#3565)

Commit:
Milestone 3 of extending autosave functionality

Commit:
Merge branch 'develop' into autosave-local

Commit:
remove unneeded comments

Commit:
remove tabs

Commit:
remove tabs

Commit:
check storage exists before removing item

seanlip commented:
Hi @Anthony-Alridge -- I took a quick look at this, and I want to ask for a general refactor. Can you make a new file for a LocalStorageService.js that abstracts away the local storage stuff, so that this service can act as the interface for all uses of localStorage, and we can do things like LocalStorageService.saveExplorationDraft()? It would be similar to siteAnalyticsService in app.js. I think that would make it easier to test and keep things conceptually clear.

I'm fine with deferring the modal thing to a future PR, and e2e tests to a PR where they make sense. I think in this PR you can include some tests for the operation of LocalStorageService once you've done the refactor in the previous paragraph.

Anthony-Alridge commented:
Hi @seanlip yes that makes sense, I'll get on that then. Thanks!

Commit:
Introduced a local storage save service for exploration drafts

Commit:
Merge branch 'develop' into autosave-local

Commit:
Fix lint errors

Commit:
fix lint error

Anthony-Alridge commented:
Hi @seanlip I've made an attempt at that refactor, please have a look. Thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -20,9 +20,10 @@
 oppia.factory('explorationData', [
   '$http', '$log', 'alertsService',
   'EditableExplorationBackendApiService',
+  'localSaveService',
```
LocalStorageService. We're naming services as UpperCamelCase (but haven't fully finished converting the old ones yet). Also, in the future, we may do other things with localStorage so it seems to make sense to give it a more general name.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -20,9 +20,10 @@
   'EditableExplorationBackendApiService',
+  'localSaveService',
   'ReadOnlyExplorationBackendApiService','$q',
   function($http, $log, alertsService, EditableExplorationBackendApiService,
```
Let's try and have the line breaks here match those in the list of strings above, so that it's easy to compare both at a glance and verify that they're the same.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -50,12 +50,16 @@ oppia.factory('explorationData', [
       explorationId: explorationId,
       autosaveChangeList: function(changeList, successCallback, errorCallback) {
+        // First save locally to be retrieved later if save is unsuccessful.
+        localSaveService.saveExplorationDraft(explorationId, changeList,
```
If the contents of (...) span multiple lines, break after the '('.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -50,12 +50,16 @@ oppia.factory('explorationData', [
-            draftChangeListId = response.data.draft_change_list_id;
             successCallback(response);
           }
         }, function() {
```
Is there a chance of an error if the draftChangeListId that's sent to the backend is old for some reason (as might happen if the user's computer loses connection and then later connects to the Internet after they've edited from another computer)? If so, would that need to be handled with the local storage service?

Anthony-Alridge replied:
I'm not entirely sure what you mean. The draftChangeListId is only sent from the back-end to the front-end (on autosaves and loads) so that shouldn't be possible. In the scenario you described the draftChangeListId on the locally saved draft would differ from that on the back-end so the it would be discarded (preferably after showing a modal displaying the changes which may be lost).  Does that address your concern or have I misunderstood? Thanks!

seanlip replied:
I think that does explain it, thanks! Do we have plans for adding that modal that displays the potentially-lost changes? If it's part of the milestones, that sounds good -- it's an important use case, I think.

Anthony-Alridge replied:
Yep I was planning to add it on the next PR which will also deal with the case where the user loses changes due to them not having access to localStorage for whatever reason. It wasn't part of the original milestones but like you said I think it's an important use-case. I'd be happy to add it to this PR if you'd prefer of course.

seanlip replied:
Oh yes, I suggested implementing milestone 4 in this PR. I think it makes sense to do this here as well, yeah. Then you can test the end-to-end flow before ever saving anything in production.

Thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -50,12 +50,16 @@ oppia.factory('explorationData', [
     var explorationData = {
       explorationId: explorationId,
       autosaveChangeList: function(changeList, successCallback, errorCallback) {
```
Can you add a comment to explain whether changeList is the incremental changelist (i.e. between draftChangeListId - 1 and the present) or whether it's the full changelist since the last saved/committed version?

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+   * Create the key to access the changeList in localStorage
+   * @param {String} key - The exploration id of the changeList to be accessed.
+   */
+   var _createLocalSaveKey = function(key) {
```
_createExplorationDraftKey (since that's what you're calling the thing you're saving)

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+   * @param {String} key - The exploration id of the changeList to be accessed.
+   */
+   var _createLocalSaveKey = function(key) {
+     return 'draft_' + key;
```
Use explorationId instead of key for the variable name. More explicit.

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+       return Boolean(storage);
+     },
+     /**
+     * Save the given changeList to localStorage along with it's
```
its

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+     * draftChangeListId
+     * @param {String} explorationId - The id of the exploration
+     *   associated with the changeList to be saved.
+     * @param {Object} changeList - The exploration change list to be saved.
```
I thought this would be an array?

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+       if (storage) {
+         var saveObject = {
+           draftChanges: changeList,
+           draftChangeListId: draftChangeListId};
```
Move } to next line.

Should the save object also include the exploration version (to help you tell whether it's obsolete or not)?

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+     getExplorationDraft: function(explorationId) {
+       if (storage) {
+         return JSON.parse(storage.getItem(_createLocalSaveKey(explorationId)));
+       }
```
else return null?

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+     * Retrieve the local save of the changeList associated with the given
+     * exploration id.
+     * @param {String} explorationId - The exploration id of the change list
+     *   to be retrieved.
```
For JSDoc, we need @returns as well, right?

seanlip commented on core/templates/dev/head/services/localSaveService.js:
```diff
@@ -0,0 +1,93 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
```
Consider updating to 2017. Ditto in the spec file.

seanlip commented on core/templates/dev/head/services/localSaveServiceSpec.js:
```diff
@@ -0,0 +1,59 @@
+     var draftChangeListIdTwo = 1;
+     var saveOne = {
+       draftChanges: changeList,
+       draftChangeListId: draftChangeListIdOne};
```
Here and below, move }; to next line.

seanlip reviewed (commented):
Thanks @Anthony-Alridge, I took a first pass!

One suggestion I have is to implement milestone 4 in this PR as well, together with tests. I feel that you'll only really know that the saving stuff works if you're able to subsequently load it properly :)

Commit:
Address review coments

Commit:
renaming local save service to local storage service.

Commit:
renaming localSaveService.js

Commit:
Milestone 4 completed

Commit:
Show lost changes modal when an invalid draft is in local storage

Commit:
offline autosave load is working and modal added.

Commit:
modify LocalStorageService tests to reflect new name

seanlip commented:
Hi @Anthony-Alridge, how's this going?

Commit:
Fix indentation errors

Anthony-Alridge commented:
Hi @seanlip thanks for asking! It's coming along well, almost finished now. I'm still having trouble with end-to-end testing however, the problem being that there doesn't seem to be a suitable way to simulate having the browser offline during automated testing which is of course needed to test this feature. I've manually tested and everything seems to be working. Do you have any idea how to approach this? At the moment I'm thinking of adding more unit tests, at the very least to check that the error modal is opened when  there is a mismatch with draft ids.

seanlip commented:
Thanks, @Anthony-Alridge! I think manual tests and more karma unit tests should be fine in this case -- I agree with you that handling this in e2e tests is hard.

Would you like a code review at this point? If not, just let me know when.

Thanks!

Anthony-Alridge commented:
Hi @seanlip, yes if you could review that would be great. Thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -18,11 +18,18 @@
 // Service for handling all interactions with the exploration editor backend.
 oppia.factory('explorationData', [
-  '$http', '$log', 'alertsService',
+  '$http', '$log',
```
Hm, in a way, having one dependency per line is neat, but it's weird for this to be the only place in the entire codebase where this happens. Could you bring it back to how it was, where we try to use most of the line width?

Also, in the process, it'd be great if you could move $q to near the beginning as well, so that all the $-dependencies come first in the list. Thanks!

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -49,13 +55,21 @@ oppia.factory('explorationData', [
     var explorationData = {
       explorationId: explorationId,
+      // Note that the changeList is the full changeList since the last
+      // commited version.
```
committed (typo)

And, maybe to make it clearer: `since the last committed version (as opposed to the most recent autosave)`.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -89,14 +104,29 @@ oppia.factory('explorationData', [
           // changes applied. This makes a force-refresh necessary when changes
           // are discarded, otherwise the exploration-with-draft-changes
           // (which is cached here) will be reused.
+          console.log(errorCallback);
```
Please do not leave console.logs in submitted code.

Anthony-Alridge replied:
Oops, sorry about that. Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -89,14 +104,29 @@ oppia.factory('explorationData', [
               draftChangeListId = response.draft_change_list_id;
               explorationData.data = response;
-
+              var draftChanges = LocalStorageService.getExplorationDraft(
```
It's confusing to see `draftChanges.draftChanges`. Maybe just call this `draft`?

Also, consider creating a domain object for ExplorationDraft with private properties `changeList` and `id`. It has methods toJson() on its prototype, and a createFromJson() method. That way you are passing around an actual domain object representing a draft, and not just a freely-mutable dict.

Anthony-Alridge replied:
Sure. I've gotten started on that, just need to edit it a little so should send another push today to finish that up.

Anthony-Alridge replied:
I've had a go at that but with slightly different methods from what you've suggested. What do you think? I wasn't sure about the toJSON method on it's prototype as I'm not sure when it will be used (as the domain object will only be converted to JSON when saving to localStorage which is the same time as when the domain object is created so we'd have something like ExplorationDraftObjectFactory.createFromJSON().toJSON()). Is that what we want or is the way it is currently implemented okay? Thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -89,14 +104,29 @@ oppia.factory('explorationData', [
+                draftChanges.draftChangeListId === draftChangeListId) {
+                var changeList = draftChanges.draftChanges;
+                explorationData.autosaveChangeList(changeList, function() {
+                  // a reload is needed so that the changelist just saved is
```
Capitalize "A" at the start; end with a period.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -2537,6 +2569,38 @@ oppia.factory('autosaveInfoModalsService', [
+              $modalInstance.dismiss('cancel');
+            };
+
+            $scope.hasLostChanges = (lostChanges && lostChanges.length > 0);
```
Not sure I understand ... why would we even show this modal if there are no lost changes?

Anthony-Alridge replied:
Ah that's my bad, copy-paste error. Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/ExplorationEditor.js:
```diff
@@ -105,7 +105,12 @@ oppia.controller('ExplorationEditor', [
     // page load.
     $scope.initExplorationPage = function(successCallback) {
-      explorationData.getData().then(function(data) {
+      explorationData.getData(function(explorationId, lostChanges){
```
nit: add space before {

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/lost_changes_modal.html:
```diff
@@ -0,0 +1,22 @@
+<div class="modal-header">
+  <h3>Error With Exploration</h3>
```
Error Saving Changes

Anthony-Alridge replied:
I've renamed this Error Loading Changes as it will be called when a user attempts to load an exploration but has draft changes stored in local storage which have been overwritten by the saved draft changes. Is that alright?

seanlip replied:
Yes, that's fine!

seanlip commented on core/templates/dev/head/pages/exploration_editor/lost_changes_modal.html:
```diff
@@ -0,0 +1,22 @@
+<div class="modal-body">
+  <p>
+    Sorry! The following changes will be lost. This is likely due to
+    editing an exploration without internet connection.
```
Can we be a bit more explicit? What exactly happened? (Was their draft overwritten, for example?)

You don't want to give the impression that editing an exploration without Internet connection messes things up automatically, if that's not accurate.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,101 @@
+ // Note that the draft is only saved if localStorage exists and works
+ // (i.e. has storage capacity).
+ oppia.factory('LocalStorageService', function() {
+   // check that local storage exists and works as expected.
```
Start comments with a capital letter.

Anthony-Alridge replied:
Done.

seanlip reviewed (changes requested):
Took a pass -- I think this looks pretty good! Thanks @Anthony-Alridge. 

PTAL at comments. For more info on domain objects, you might want to look at #2863.

Commit:
Address review comments and add exploration draft object factory

Commit:
Add spec for draft object factory

Commit:
Load new object factory in exploration_editor.html

Commit:
fix linting errors

Commit:
fix linting errors

Commit:
fix typo

Commit:
rename method createFromDict to createFromJSON

Commit:
update tests to reflect new method names

Anthony-Alridge commented:
Hi @seanlip I've addressed your review comments, please have a look. Thanks! Also, unfortunately I'm still having trouble testing the feature, in particular the issue at the moment is in testing explorationData. The problem is that to instantiate the explorationData instance the exploration_id has to be recovered from the window.location.pathname . I'm not entirely sure how to get around this properly, an option is refactoring out the process into a service retrieving the exploration id from the pathname so that this service can then be mocked. Is it worth exploring this approach or would manual testing suffice?

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactory.js:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * @fileoverview Factory for creating instances of ExplorationDraft
+ * domain objects..
```
Drop extra period at end.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactory.js:
```diff
@@ -0,0 +1,43 @@
+       this.draftChangeListId = draftChangeListId;
+     };
+
+     ExplorationDraft.prototype.isDraftValid = function(validDraftId) {
```
Maybe a better name is currentDraftId? I suggest adding a brief docstring here to explain the context (what does a "valid draft" mean) and the meaning of the arg that's being passed in.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactory.js:
```diff
@@ -0,0 +1,43 @@
+       return this.draftChanges;
+     };
+
+     ExplorationDraft.createFromJSON = function(
```
I think currently our convention in the frontend is to treat JSON as one word, so maybe call this createFromJson. if you like, fromJson() is also fine.

That said, this method isn't actually creating from JSON. JSON is an encoding format for strings, but what you're passing in here is clearly a dict (or JS object). If that's the case maybe you can call it createFromLocalStorageDict instead.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactorySpec.js:
```diff
@@ -0,0 +1,48 @@
+     var explorationIdOne = '100';
+     var draftChangeListIdOne = 2;
+     var changeList = [];
+     var saveOne = {
```
Why not call this draftDict or draftDictOne? (if you want to distinguish from the draftObject? Though normally we just call the draftObject a "draft").

You probably don't need the "One" all over the place, actually, since there's just one thing here.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -96,7 +105,21 @@ oppia.factory('explorationData', [
+              var draft = LocalStorageService.getExplorationDraft(
+                explorationId);
+              if (draft &&
+                draft.isDraftValid(draftChangeListId)) {
```
I think this can probably go to the end of the previous line?

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -96,7 +105,21 @@ oppia.factory('explorationData', [
+                  // response.
+                  $window.location.reload();
+                });
+              } else if (draft &&
```
Better to do:

    if (draft) {
      if (draft.isDraftValid(draftChangeListId)) {
        ...
      } else {
        ...
      }
    }

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+         draftChangeListId) {
+         var localSaveKey = _createExplorationDraftKey(explorationId);
+         if (storage) {
+           var saveObject = {
```
Let's call this draftDict.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+             draftChanges: changeList,
+             draftChangeListId: draftChangeListId
+           };
+           saveObject = JSON.stringify(saveObject);
```
Don't rewrite vars; hard to tell what a var represents that way. Just do storage.setItem(localSaveKey, JSON.stringify(draftDict));

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+       */
+       getExplorationDraft: function(explorationId) {
+         if (storage) {
+           var saveObject = JSON.parse(
```
Ditto -- draftDict

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageServiceSpec.js:
```diff
@@ -0,0 +1,67 @@
+     var changeList = [];
+     var explorationIdTwo = '101';
+     var draftChangeListIdTwo = 1;
+     var saveOne = {
```
Ditto re naming.

Anthony-Alridge replied:
Done.

seanlip reviewed (changes requested):
Thanks @Anthony-Alridge! Took another pass.

seanlip commented:
> Also, unfortunately I'm still having trouble testing the feature, in particular the issue at the moment is in testing explorationData. The problem is that to instantiate the explorationData instance the exploration_id has to be recovered from the window.location.pathname . I'm not entirely sure how to get around this properly, an option is refactoring out the process into a service retrieving the exploration id from the pathname so that this service can then be mocked. Is it worth exploring this approach or would manual testing suffice?

Hmm, I think we really do need to write tests for this feature. It's a bit fragile, and I can see this leading to hard-to-reproduce bugs if it's not protected with tests.

I think your suggestion is along the right lines. There is actually a urlService in app.js. Can we perhaps use that instead? Feel free to do that in a separate PR that you merge into develop before this one, if you like.

Thanks!

Commit:
Address review comments

Commit:
Modify explorationData to use urlService.getPathname() to allow for easier mocking in unit tests

Commit:
Add tests for exploartion data getData to ensure localsave draft is correctly used.

Commit:
Fix lint errors

Commit:
Fix lint errors

Commit:
Fix lint errors

seanlip commented:
Hi @Anthony-Alridge -- just dropping a quick note to say, please let me know explicitly (by writing "@seanlip PTAL") when you want a review! I'm going to assume for the moment that you haven't finished yet.

(Btw, you might want to do this by the weekend, since I'm going to be away next week :P)

Anthony-Alridge commented:
Hi @seanlip, I'm gonna give it a look over before I request a review but I should be able to ask soon. I'll tell you when. Thanks!

Commit:
remove console.log statement

Anthony-Alridge commented:
@seanlip PTAL. Thanks!

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactorySpec.js:
```diff
@@ -0,0 +1,52 @@
+     beforeEach(inject(function($injector) {
+       ExplorationDraftObjectFactory = $injector.get(
+         'ExplorationDraftObjectFactory');
+       draft =
```
Indentation:
```
draft = (
  ExplorationDraftObjectFactory.createFromLocalStorageDict(
    draftDict));
```

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactory.js:
```diff
@@ -0,0 +1,53 @@
+     * @returns {Boolean} - True iff the currentDraftId is the same as the
+     * draftChangeListId corresponding to this draft.
+     */
+     ExplorationDraft.prototype.isDraftValid = function(currentDraftId) {
```
Just a thought -- you might consider calling this isValid() and the following method getChanges(), since then you have draft.isValid() and draft.getChanges(), which reads more naturally than draft.isDraftValid() etc.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServicesSpec.js:
```diff
@@ -17,6 +17,106 @@
+        draft_change_list_id: 3,
+      };
+
+      draftChangesValid = {
```
responseWhenDraftChangesAreValid

(There should be a noun describing what the thing "is" in the name of the object.)

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/EditorServicesSpec.js:
```diff
@@ -17,6 +17,106 @@
+        }
+      };
+
+      draftChangesInvalid = {
```
responseWhenDraftChangesAreInvalid

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+         return result && localStorage;
+       } catch (exception) {}
+     }());
+     /**
```
Nit: can you add a blank line above this for readability?

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+     */
+     var _createExplorationDraftKey = function(explorationId) {
+       return 'draft_' + explorationId;
+     }
```
Nit: add semicolon at end; add blank line below this.

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+       * @param {List} changeList - The exploration change list to be saved.
+       * @param {Integer} draftChangeListId - The id of the draft to be saved.
+       */
+       saveExplorationDraft: function(explorationId, changeList,
```
When contents of parens span more than one line, break after '('. You can indent the following line by an additional two spaces, too, to distinguish its left edge from the stuff that's actually inside the (...).

Anthony-Alridge replied:
Done.

seanlip commented on core/templates/dev/head/services/LocalStorageService.js:
```diff
@@ -0,0 +1,105 @@
+         draftChangeListId) {
+         var localSaveKey = _createExplorationDraftKey(explorationId);
+         if (storage) {
+           var draftDict = {
```
Maybe here it makes sense to do ExplorationDraftObjectFactory.toLocalStorageDict()?

Anthony-Alridge replied:
Done.

seanlip reviewed (commented):
Hi @Anthony-Alridge -- on the whole, this looks great, and I cannot see any issues! I just have some trivial code comments to address before merge, but otherwise it looks fantastic. Thank you very much!

Commit:
Addressing review comments.

Commit:
Address review comments.

Commit:
Fix lint errors

Commit:
Address review comments

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationDraftObjectFactory.js:
```diff
@@ -0,0 +1,62 @@
+         explorationDraftDict.draftChangeListId);
+     };
+
+     ExplorationDraft.toLocalStorageDict = function(
```
Add Karma test for this, please.

Anthony-Alridge replied:
Done.

Anthony-Alridge commented:
@seanlip PTAL

seanlip reviewed (approved):
LGTM, thanks @Anthony-Alridge! Please add the missing Karma test, and manually test that the e2e flow is working correctly in all cases. Then it should be good to go!

Commit:
Add karma test for exploration draft object to local storage dict.

Anthony-Alridge commented:
@seanlip I've added the karma test and did the manual testing. Thanks!

seanlip commented:
Great, thanks! LGTM. I've restarted the travis tests.

Commit:
Merge branch 'develop' into autosave-local

Commit:
Fix missing script load causing collection tests to fail

Commit:
Fix missing script load causing collection tests to fail

tjiang11 commented:
Is this ready to go?

seanlip commented:
It should be, I think. I've restarted the Travis test, let's merge this when the tests pass.
